### PR TITLE
Basic Spout implementation

### PR DESCRIPTION
**Pull request by adelelopez**
This creates a basic implementation of spouts as described here: https://github.com/pachyderm/pachyderm/issues/2958

## How to create and use a spout
Spouts are created in a similar way to services, and include an overwrite flag to toggle whether the spout overwrites the output file or appends to it. The PipelineInfo struct now has a new field, `Spout` which takes a pointer to a Spout struct.

The spout pipeline should run something that spouts data, writes the data to files, and then puts those files in a tar archive in `/pfs/out`.  

## How spouts work
The output file is a named pipe, and the spout asynchronously runs a receiver that continuously opens the pipe, reads in new files from it (formatted as a tar archive), commits the new files, and then closes the pipe.

The spout receiver will initially block when it opens the pipe. Meanwhile, the spout pipeline will write new data to the pipe. Once the write writes an end of file, the receiver will be able to finish opening the pipe, and it can commit those files in. Then it will start the process again.

## Other changes made
I also refactored the code for pipeline masters, unifying them into a single master function which calls the appropriate spawner.

I also took out the in the masters to mark pipelines as running, because `monitorPipelines` takes care of this already. I verified that pipelines still were marked as running as expected after taking this out.

### CODE REVIEW

File: src/server/worker/master.go [hunk 1]
```diff
@@ -117,31 +97,8 @@ func (a *APIServer) serviceMaster() {
 			return err
 		}
 		defer masterLock.Unlock(ctx)
-
 		logger.Logf("Launching master process")
```
Comment (adelelopez) [inline src/server/worker/master.go]: One thing I might want to change: we don't have a way to log which spawner this master process is using, which might be a good thing to log for debugging purposes. I think the simplest and most robust way would just be to add a `spawnerName` field to the master function. Trying to hacky things to get the function name directly from the spawner pointer might also be possible, but seems like it would add undue complexity here.

File: src/server/pfs/server/driver.go [hunk 2]
```diff
@@ -1346,7 +1354,7 @@ func (d *driver) deleteCommit(pachClient *client.APIClient, userCommit *pfs.Comm
 		}
 
 		// 3) Validate the commit (check that it has no provenance) and delete it
-		if len(userCommitInfo.Provenance) > 0 {
+		if provenantOnInput(userCommitInfo.Provenance) {
```
Comment (adelelopez) [inline src/server/pfs/server/driver.go]: Just want to double check that this is doing the right thing here.

Comment (jdoliner) [reply src/server/pfs/server/driver.go]: So doing this will make it so that people can delete commits that were emitted by a spout pipeline, which seems to make sense, since they're more like input commits than output commits.

It would be good to have a test where we delete a commit that was outputted by a spout to make sure this behavior works as we expect and doesn't break in the future.

File: src/server/pachyderm_test.go [hunk 3]
```diff
@@ -8285,3 +8285,184 @@ func getEtcdClient(t testing.TB) *etcd.Client {
...
+				// this time, we expect our commits to have different provenance from the commits earlier
+				require.NotEqual(t, provenanceID, provenance.ID)
+				provenanceID = provenance.ID
+			} else {
+				// but they should still have the same provenance as each other
+				require.Equal(t, provenanceID, provenance.ID)
+			}
+		}
+	})
+}
```
Comment (jdoliner) [inline src/server/pachyderm_test.go]: Could we also have a test that attaches a downstream pipeline to the spout. I have no reason to suspect that won't work but that's going to be the main use case for spouts so it'd be good to have some test coverage on it.

File: src/server/pps/pretty/pretty.go [hunk 4]
```diff
@@ -368,6 +368,8 @@ func prettyTransform(transform *ppsclient.Transform) (string, error) {
 // ShorthandInput renders a pps.Input as a short, readable string
 func ShorthandInput(input *ppsclient.Input) string {
 	switch {
+	case input == nil:
+		return ""
```
Comment (jdoliner) [inline src/server/pps/pretty/pretty.go]: Let's do something a little prettier than an empty string here, maybe we could do `"none"`.

File: src/server/pps/server/api_server.go [hunk 5]
```diff
@@ -2309,15 +2313,13 @@ func (a *apiServer) deletePipeline(pachClient *client.APIClient, request *pps.De
...
-			}
-		})
-	}
+	pps.VisitInput(pipelineInfo.Input, func(input *pps.Input) {
+		if input.Cron != nil {
+			eg.Go(func() error {
+				return pachClient.DeleteRepo(input.Cron.Repo, request.Force)
+			})
+		}
+	})
```
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: This is an unrelated improvement you're able to make now because `VisitInput` correctly handles `nil` right? If so that's a nice bonus, I wonder why we were already defending against `nil` inputs here, I guess at one point we actually allowed them.

Comment (adelelopez) [reply src/server/pps/server/api_server.go]: Yep, that's right

File: src/server/worker/master.go [hunk 6]
```diff
@@ -72,7 +74,7 @@ func (logger *taggedLogger) jobLogger(jobID string) *taggedLogger {
 	return result
 }
 
-func (a *APIServer) master() {
+func (a *APIServer) master(spawnerName string, spawner func(*client.APIClient) error) {
```
Comment (jdoliner) [inline src/server/worker/master.go]: Passing in a string here is a great way to solve the problem. Minor nits:
- could we call the variable `masterType` that seems to match up with how it's use a bit better.
- for the values passed in could we have them be: `"service"`, `"spout"` and `"pipeline"`, I think that will lead to the most understandable logs for users.

File: src/server/pachyderm_test.go [hunk 7]
```diff
@@ -8285,3 +8285,200 @@ func getEtcdClient(t testing.TB) *etcd.Client {
...
+		require.NoError(t, c.CreatePipeline(
+			pipeline2,
+			"",
+			[]string{"/bin/bash"},
+			[]string{"cp " + fmt.Sprintf("/pfs/%s/*", pipeline) + " /pfs/out/"},
+			nil,
+			client.NewPFSInput(pipeline, "/*"),
+			"",
+			false,
+		))
```
Comment (jdoliner) [inline src/server/pachyderm_test.go]: One last thing, to make sure that this pipeline is actually functioning properly could you do a `FlushJobAll` and check that the jobs are succeeding. You can see an example on line `1231` in this file.
